selfdrive/ui/qt/sidebar.cc

```diff
@@ -3,185 +3,110 @@
-SignalWidget::SignalWidget(QWidget *parent) : QFrame(parent), _strength(0) {
-  layout = new QVBoxLayout();
-  layout->setMargin(0);
-  layout->setSpacing(0);
-  layout->insertSpacing(0, 45);
-
-  label = new QLabel(this);
-  label->setStyleSheet(R"(font-size: 35px; font-weight: 400;)");
-  layout->addWidget(label, 0, Qt::AlignLeft);
-
-  setFixedWidth(177);
-  setLayout(layout);
 }
 
-void SignalWidget::paintEvent(QPaintEvent *e) {
-  QPainter p(this);
-  p.setRenderHint(QPainter::Antialiasing, true);
-  p.setPen(Qt::NoPen);
-  p.setBrush(Qt::white);
-  for (int i = 0; i < 5; i++) {
-    if (i == _strength) {
-      p.setPen(Qt::NoPen);
-      p.setBrush(Qt::darkGray);
+void Sidebar::update(const UIState &s) {
+  if (s.sm->frame % (6*UI_FREQ) == 0) {
+    connect_str = "OFFLINE";
+    connect_status = warning_color;
+    auto last_ping = params.get<float>("LastAthenaPingTime");
+    if (last_ping) {
+      bool online = nanos_since_boot() - *last_ping < 70e9;
+      connect_str = online ? "ONLINE" : "ERROR";
+      connect_status = online ? good_color : danger_color;
     }
-    p.drawEllipse(QRectF(_dotspace * i, _top, _dia, _dia));
   }
-}
 
-void SignalWidget::update(const QString &text, int strength) {
-  label->setText(text);
-  if (_strength != strength) {
-    _strength = strength;
-    repaint();
+  temp_status = danger_color;
+  auto ts = s.scene.deviceState.getThermalStatus();
+  if (ts == cereal::DeviceState::ThermalStatus::GREEN) {
+    temp_status = good_color;
+  } else if (ts == cereal::DeviceState::ThermalStatus::YELLOW) {
+    temp_status = warning_color;
   }
-}
-
-Sidebar::Sidebar(QWidget *parent) : QFrame(parent) {
-  QVBoxLayout *layout = new QVBoxLayout();
-  layout->setContentsMargins(25, 50, 25, 50);
-  layout->setSpacing(35);
-  setFixedSize(300, vwp_h);
-
-  QPushButton *s_btn = new QPushButton;
-  s_btn->setStyleSheet(R"(
-    border-image: url(../assets/images/button_settings.png);
-  )");
-  s_btn->setFixedSize(200, 117);
-  layout->addWidget(s_btn, 0, Qt::AlignHCenter);
-  QObject::connect(s_btn, &QPushButton::pressed, this, &Sidebar::openSettings);
-
-  signal = new SignalWidget(this);
-  layout->addWidget(signal, 0, Qt::AlignTop | Qt::AlignHCenter);
-
-  temp = new StatusWidget(true, this);
-  layout->addWidget(temp, 0, Qt::AlignTop);
-
-  panda = new StatusWidget(false, this);
-  layout->addWidget(panda, 0, Qt::AlignTop);
-
-  connect = new StatusWidget(false, this);
-  layout->addWidget(connect, 0, Qt::AlignTop);
-
-  QImage image = QImageReader("../assets/images/button_home.png").read();
-  QLabel *comma = new QLabel(this);
-  comma->setPixmap(QPixmap::fromImage(image));
-  comma->setAlignment(Qt::AlignCenter);
-  layout->addWidget(comma, 1, Qt::AlignHCenter | Qt::AlignVCenter);
-
-  layout->addStretch(1);
-
-  setStyleSheet(R"(
-    Sidebar {
-      background-color: #393939;
-    }
-    * {
-      color: white;
-    }
-  )");
-  setLayout(layout);
-}
+  temp_val = (int)s.scene.deviceState.getAmbientTempC();
```

Comment:
@deanlee can we fix up the C++ submaster so we always have a valid reader like the python version? then we can remove all these extra readers in UI.

Reply:
good idea

Reply:
https://github.com/commaai/cereal/pull/139